docs/proposals/stateful-failover/stateful-failover-proposal.md

```diff
@@ -0,0 +1,197 @@
+type ResourceBindingStatus struct {
+	// SchedulerObservedGeneration is the generation(.metadata.generation) observed by the scheduler.
+	// If SchedulerObservedGeneration is less than the generation in metadata means the scheduler hasn't confirmed
+	// the scheduling result or hasn't done the schedule yet.
+	// +optional
+	SchedulerObservedGeneration int64 `json:"schedulerObservedGeneration,omitempty"`
+
+	// SchedulerObservedAffinityName is the name of affinity term that is
+	// the basis of current scheduling.
+	// +optional
+	SchedulerObservedAffinityName string `json:"schedulerObservingAffinityName,omitempty"`
+
+	// Conditions contain the different condition statuses.
+	// +optional
+	Conditions []metav1.Condition `json:"conditions,omitempty"`
+
+	// AggregatedStatus represents status list of the resource running in each member cluster.
+	// +optional
+	AggregatedStatus []AggregatedStatusItem `json:"aggregatedStatus,omitempty"`
+
+    // FailoverHistory represents history of the previous failovers of this resource
+	FailoverHistory []FailoverHistoryItem `json:"failoverHistory,omitempty"`
+
+}
+```
+
+The FailoverHistory is a list of FailoverHistoryItem objects and is updated every time a failover happens until a certain limit. The limit is set by another field persistedFields.maxHistory which is defined in the propogation policy.
+
+```
+type FailoverHistoryItem struct {
+
+    // FailoverTime represents the timestamp when the workload failed over
+    // It is represented in RFC3339 form(like '2021-04-25T10:02:10Z') and is in UTC
+	FailoverTime *metav1.Time `json:"failoverTime,omitempty"`
+
+    // OriginCluster represents the cluster name from which the workload failed over from
+	OriginCluster string `json:"originCluster,omitempty"`
+
+    // DestinationCluster represents the cluster name from which the workload failed over to
+	DestinationCluster string `json:"destinationCluster,omitempty"`
+
+    // PersistedDuringFailover contains the fields required by the stateful application to resume from that state after failover
+	PersistedDuringFailover []PersistedFailoverItem `json:"persistedFailoverItem,omitempty"` 
+}
+```
+
+The FailoverHistoryItem object contains information relevant to a failover and an additional object called "PersistedDuringFailover" which keeps track of the metadata (both the fields that need to be persisted and how to access those fields) that is required by the stateful operation to resume processing from that state.
+
+```
+type PersistedDuringFailover struct {
+	
+    // LabelName represents the name of the line that will be persisted for the replica
+    // in case there is a failover to a new cluster.
+	LabelName string `json:"labelName,omitempty"`
+
+    // PersistedItem is a pointer to the status item that should be persisted to the rescheduled 
+    // replica during a failover. This should be input in the form: obj.status.<path-to-item> 
+	PersistedStatusItem string `json:"persistedStatusItem,omitempty"`
```

Comment:
Given this `persistedStatusItem` will be applied as a label value, which brings a limitation that the length can not exceed 63 characters. Does Flink's checkpoint path exceed this?

Reply:
I should make a change here:

> // persistedStatusItem is ~~a pointer to the status item~~ the value of the status item that should be persisted to the rescheduled replica during a failover. This should be input in the form: ~~obj.status.path-to-item~~ value at the location obj.status.path-to-item

In the case of flink, the flink process will create a jobID at obj.status.jobStatus.jobID which would look like `fd72014d4c864993a2e5a9287b4a9c5d` so the persistedStatusItem would store this value in the PersistedStatusItem field.

Flink Deployment spec would have:
```
status:
  jobStatus:
    jobID: fd72014d4c864993a2e5a9287b4a9c5d
```

PersistedDuringFailover would store jobID as the labelName and fd72014d4c864993a2e5a9287b4a9c5d as the PersistedStatusItem (value)

Reply:
Since we are storing the metadata related to the state here, I think the limitation of the value being a string and less than 63 characters is okay. The actual state of the flink job will be stored in some durable storage and may be very big. In other stateful applications as well I think we can make this assumption that metadata related to the state should be persisted with the resourceBindingStatus and the actual data can be stored elsewhere.

I will include this in the proposal. Let me know what you think

Reply:
Looks like as you said annotations values can support 253 characters whereas label values can only be 64 characters long. Annotation keys can only be 64 characters whereas label keys can be 253 characters.